### PR TITLE
[mlir] Method to iterate over registered operations for a given dialect class.

### PR DESCRIPTION
**Pull request by Rajveer100**
Part of #111591

Currently we have `MLIRContext::getRegisteredOperations` which returns all operations for the given context, with the addition of `MLIRContext::getRegisteredOperationsByDialect` we can now retrieve the same for a given dialect class.

### CODE REVIEW

**Comment by Rajveer100** [comment]
@stellaraccident

**Comment by Rajveer100** [comment]
Before proceeding with comments and test cases, let me know if this works well.

File: mlir/include/mlir/IR/MLIRContext.h [hunk 1]
```diff
@@ -195,7 +195,12 @@ class MLIRContext {
 
   /// Return a sorted array containing the information about all registered
   /// operations.
-  ArrayRef<RegisteredOperationName> getRegisteredOperations();
+  SmallVector<RegisteredOperationName, 0> getRegisteredOperations();
```
Comment (joker-eph) [inline mlir/include/mlir/IR/MLIRContext.h:198]: It's not clear to me why we can't just keep returning an ArrayRef?

A copy into a vector makes it quite expensive.

File: mlir/lib/IR/MLIRContext.cpp [hunk 2]
```diff
@@ -188,7 +188,11 @@ class MLIRContextImpl {
 
   /// This is a sorted container of registered operations for a deterministic
   /// and efficient `getRegisteredOperations` implementation.
-  SmallVector<RegisteredOperationName, 0> sortedRegisteredOperations;
+  SmallVector<std::pair<StringRef, RegisteredOperationName>, 0>
```
Comment (joker-eph) [inline mlir/lib/IR/MLIRContext.cpp:191]: I don't follow why we need this StringRef here. The dialect name is already available from the RegisteredOperationName, isn't it?

**Comment by stellaraccident** [comment]
Per usual, Mehdi beat me to the review. Thank you for the PR.

File: mlir/lib/Rewrite/FrozenRewritePatternSet.cpp [hunk 3]
```diff
@@ -73,7 +73,7 @@ FrozenRewritePatternSet::FrozenRewritePatternSet(
   // Functor used to walk all of the operations registered in the context. This
   // is useful for patterns that get applied to multiple operations, such as
   // interface and trait based patterns.
-  std::vector<RegisteredOperationName> opInfos;
+  SmallVector<RegisteredOperationName> opInfos;
```
Comment (stellaraccident) [inline mlir/lib/Rewrite/FrozenRewritePatternSet.cpp:76]: I'm not immediately seeing why this change is related to the rest of the PR (although probably a good change in general, I would expect)?

File: mlir/include/mlir/IR/MLIRContext.h (hunk 1 above)
Comment (Rajveer100) [reply mlir/include/mlir/IR/MLIRContext.h:198]: Yes, I had this in my mind, with the latest change I have added a member variable so we can continue to use `ArrayRef` whilst using `transform`.

File: mlir/lib/Rewrite/FrozenRewritePatternSet.cpp (hunk 3 above)
Comment (Rajveer100) [reply mlir/lib/Rewrite/FrozenRewritePatternSet.cpp:76]: Reverted this, since now we continue to use `ArrayRef` as the return type. I had initially changed the type to `SmallVector` which caused the old logic to throw an error for `=`.

File: mlir/lib/IR/MLIRContext.cpp (hunk 2 above)
Comment (joker-eph) [reply mlir/lib/IR/MLIRContext.cpp:191]: I don't quite understand what you mean here? I probably should checkout the PR locally and build it to figure out myself the problem.

Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp:191]: Essentially, as you can notice we have pairs (D1, O1), (D2, O2)... and so on (where D=DialectName and O=OperationName).

Now, while inserting I am getting the lower bound for (Di, Oi) which is the next pair to insert so it gets fit in the right place in the segment of operations for a specific dialect, i.e, (D, O1), (D, O2), (D, Oi), here all D's are the same.

For retrieval, we do it in a similar way.

If we don't use pairwise elements we wouldn't know which segment (dialect class) of the sorted elements list the operation goes to since we only have the count of operations for each dialect class (via the string map I added), but we need at least one index, either start or end.

File: mlir/lib/IR/MLIRContext.cpp [hunk 4]
```diff
@@ -188,7 +188,15 @@ class MLIRContextImpl {
...
+      sortedRegisteredOperations;
+
+  /// This stores the transformed operations when calling
+  /// `getRegisteredOperations`.
+  SmallVector<RegisteredOperationName, 0> transformedOperations;
+
+  /// This returns the number of registered operations for a given dialect.
+  llvm::DenseMap<StringRef, size_t> getCountByDialectName;
```
Comment (River707) [inline mlir/lib/IR/MLIRContext.cpp:199]: Why not have this stored in the dialect itself? That'd remove an extra map.

Also, nit: please drop the llvm:: (DenseMap is already exported in the mlir namespace).

File: mlir/lib/IR/MLIRContext.cpp (hunk 2 above)
Comment (River707) [reply mlir/lib/IR/MLIRContext.cpp:191]: I don't see why you need the dialect name stored this way, there are many ways around it. The easiest would be getting the dialect name from the RegisteredOperationName for comparison in the lower bound: `getDialect().getNamespace()`. (There are other ways as well, e.g. comparing a string directly against the operation name). That would remove the need to have the StringRef here, or the `transformedOperations` field.

**Review by River707 — CHANGES_REQUESTED** [review]

File: mlir/lib/IR/MLIRContext.cpp (hunk 2 above)
Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp:191]: Yeah, that's much better, I missed it :)

File: mlir/lib/IR/MLIRContext.cpp [hunk 5]
```diff
@@ -976,12 +994,19 @@ void RegisteredOperationName::insert(
...
       llvm::upper_bound(ctxImpl.sortedRegisteredOperations, value,
                         [](auto &lhs, auto &rhs) {
-                          return lhs.getIdentifier().compare(
-                              rhs.getIdentifier());
+                          if (lhs.getDialect().getNamespace() ==
+                              rhs.getDialect().getNamespace())
+                            return lhs.getIdentifier().compare(
+                                rhs.getIdentifier());
+                          return lhs.getDialect().getNamespace().compare(
+                              rhs.getDialect().getNamespace());
```
Comment (joker-eph) [inline mlir/lib/IR/MLIRContext.cpp:1009]: I'm confused: what are you trying to achieve here? Why isn't the existing code good enough?

File: mlir/lib/IR/MLIRContext.cpp (hunk 4 above)
Comment (joker-eph) [reply mlir/lib/IR/MLIRContext.cpp]: This isn't done right now?
The suggestion I believe is to add a "int operationCount = 0;` in the `class Dialect` where the counter would live and remove this map entirely from the Context.

File: mlir/lib/IR/MLIRContext.cpp (hunk 5 above)
Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp:1009]: https://github.com/llvm/llvm-project/pull/112344#discussion_r1803450430

So that we have segments ordered correctly when we call `getRegisteredOperationsByDialect`, the existing code won't work for this.

Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp:1009]: Sorry, misunderstood, I just realised that dialect is common here, so we don't need this? I thought the same for the count variable too, so introduced the map.

Comment (joker-eph) [reply mlir/lib/IR/MLIRContext.cpp:1009]: Can you walk me through a concrete example where the existing code does not work and this is needed?

File: mlir/lib/IR/MLIRContext.cpp (hunk 4 above)
Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp]: Fixed.

File: mlir/lib/IR/MLIRContext.cpp (hunk 5 above)
Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp]: This existing code would indeed work.

**Comment by Rajveer100** [comment]
@joker-eph 
I have made the changes, I think it should make sense now.

File: mlir/lib/IR/MLIRContext.cpp [hunk 6]
```diff
@@ -711,6 +714,27 @@ ArrayRef<RegisteredOperationName> MLIRContext::getRegisteredOperations() {
...
+                         return lhs.getDialect().getNamespace().compare(
+                             rhs.getDialect().getNamespace());
+                       });
+
+  auto upperBound =
+      std::lower_bound(lowerBound, impl->sortedRegisteredOperations.end(),
+                       dialectName, [](auto &lhs, auto &rhs) {
+                         return lhs.getDialect().getNamespace().compare(
+                             rhs.getDialect().getNamespace());
+                       });
```
Comment (River707) [inline mlir/lib/IR/MLIRContext.cpp:733]: You shouldn't need to compute the upper bound here, you should be able to use the operationCount that is computed in the dialect. That can be grabbed from the first operation of the lowerBound (after checking it's dialect is the same as the input, to account for no operations for the dialect).

Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp:733]: In the case of no operations is returning `.end()` fine?

File: mlir/lib/IR/MLIRContext.cpp [hunk 7]
```diff
@@ -190,6 +190,9 @@ class MLIRContextImpl {
   /// and efficient `getRegisteredOperations` implementation.
   SmallVector<RegisteredOperationName, 0> sortedRegisteredOperations;
 
+  /// This returns the number of registered operations for the given dialect.
+  size_t operationCount = 0;
+
```
Comment (joker-eph) [inline mlir/lib/IR/MLIRContext.cpp:195]: The comment says that this count is "for the given dialect", but we're in the context class, so it seems to me that you implemented a global count that isn't dialect specific.

The suggestion was to put this counter in `class Dialect {...}` instead I believe

File: mlir/lib/IR/MLIRContext.cpp [hunk 8]
```diff
@@ -711,6 +714,26 @@ ArrayRef<RegisteredOperationName> MLIRContext::getRegisteredOperations() {
...
+                         return lhs.getDialect().getNamespace().compare(
+                             rhs.getDialect().getNamespace());
+                       });
+
+  if (lowerBound == impl->sortedRegisteredOperations.end() ||
+      lowerBound->getDialect().getNamespace() != dialectName)
+    return ArrayRef<RegisteredOperationName>();
+
+  size_t count =
+      lowerBound->getDialect().getContext()->getImpl().operationCount;
```
Comment (joker-eph) [inline mlir/lib/IR/MLIRContext.cpp:733]: `lowerBound->getDialect().getContext()` will return `this` pointer here: there is a single MLIRContext  object in flight.

File: mlir/lib/IR/MLIRContext.cpp (hunk 6 above)
Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp]: Actually, if we do this we don't need any extra count variable, so isn't this better?

File: mlir/lib/IR/MLIRContext.cpp (hunk 7 above)
Comment (Rajveer100) [reply mlir/lib/IR/MLIRContext.cpp]: I have updated the logic that doesn't require the count now.

**Review by joker-eph — APPROVED** [review]
LG, thanks.

**Comment by Rajveer100** [comment]
Thanks for the approval, could you land this for me?!
